src/librustc/ty/instance.rs

```diff
@@ -274,13 +274,24 @@ impl<'a, 'b, 'tcx> Instance<'tcx> {
                 _ => {
                     if Some(def_id) == tcx.lang_items().drop_in_place_fn() {
                         let ty = substs.type_at(0);
-                        if ty.needs_drop(tcx, ty::ParamEnv::reveal_all()) {
-                            debug!(" => nontrivial drop glue");
-                            ty::InstanceDef::DropGlue(def_id, Some(ty))
-                        } else {
-                            debug!(" => trivial drop glue");
-                            ty::InstanceDef::DropGlue(def_id, None)
-                        }
+                        let mut def = None;
+                        if let ty::Adt(inner_def, _) = ty.sty {
+                            let va_list_did = tcx.lang_items().va_list().unwrap();
+                            if inner_def.did == va_list_did {
+                                // Resolve `drop_in_place::<VaListImpl>` to `drop_va_list`
```

Comment:
This is the ugly part.

Reply:
Please correct me if I am wrong, but I personally don't think guaranteeing that `drop_in_place::<VaListImpl>` will be optimized to a single call to `va_end` is worth this. The non-inlined version still works fine, so essentially all of this is just to ensure that the codegen test passes right? IMO we're better off either removing `inline(always)` from the `Drop` impl, or adding some extra options to the test.

Reply:
I'd really prefer to keep `inline(always)` there, since LLVM expects the `va_end` to be called from the same function as the corresponding `va_start`/`va_copy`, and always-inlining guarantees this for most cases.

Reply:
> The non-inlined version still works fine, so essentially all of this is just to ensure that the codegen test passes right?

Basically, yes. I could fix that test instead, I don't have a strong preference either way.